src/plugins/chart_expressions/expression_xy/public/helpers/color_assignment.ts

```diff
@@ -6,29 +6,86 @@
  * Side Public License, v 1.
  */
 
-import { uniq, mapValues } from 'lodash';
+import { mapValues } from 'lodash';
+import { Datatable, DatatableRow } from '@kbn/expressions-plugin';
 import { euiLightVars } from '@kbn/ui-theme';
-import { getAccessorByDimension } from '@kbn/visualizations-plugin/common/utils';
+import {
+  getAccessorByDimension,
+  getColumnByAccessor,
+} from '@kbn/visualizations-plugin/common/utils';
+import type { ExpressionValueVisDimension } from '@kbn/visualizations-plugin/common';
 import { FormatFactory } from '../types';
 import { isDataLayer } from './visualization';
 import { CommonXYDataLayerConfig, CommonXYLayerConfig } from '../../common';
 import { getFormat } from './format';
-
-const isPrimitive = (value: unknown): boolean => value != null && typeof value !== 'object';
+import { LayerAccessorsTitles } from './layers';
 
 export const defaultReferenceLineColor = euiLightVars.euiColorDarkShade;
 
 export type ColorAssignments = Record<
   string,
   {
     totalSeriesCount: number;
-    getRank(sortedLayer: CommonXYDataLayerConfig, seriesKey: string, yAccessor: string): number;
+    getRank(sortedLayer: CommonXYDataLayerConfig, seriesName: string): number;
   }
 >;
 
+function getSplitName(
+  splitAccessors: Array<ExpressionValueVisDimension | string> = [],
+  table: Datatable,
+  row: DatatableRow,
+  formatFactory: FormatFactory
+) {
+  return splitAccessors.reduce<string>((splitName, accessor) => {
+    const splitColumn = getColumnByAccessor(accessor, table.columns);
+    if (!splitColumn) return;
+    const splitAccessor = getAccessorByDimension(accessor, table.columns);
+    const columnFormatter = splitAccessor && formatFactory(getFormat(table.columns, splitAccessor));
+    const name = columnFormatter ? columnFormatter.convert(row[splitAccessor]) : row[splitAccessor];
+    if (splitName) {
+      return `${splitName} - ${name}`;
+    } else {
+      return name;
+    }
+  }, '');
+}
+
+export const getAllSeries = (
+  table: Datatable,
+  splitAccessors: CommonXYDataLayerConfig['splitAccessors'] = [],
+  accessors: Array<ExpressionValueVisDimension | string>,
+  columnToLabel: CommonXYDataLayerConfig['columnToLabel'],
+  titles: LayerAccessorsTitles,
+  formatFactory: FormatFactory
```

Comment:
Same here - let's calculate it on the top level and pass down the formatters

Reply:
done